internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -242,6 +256,61 @@ func (r *BMACReconciler) reconcileAgentSpec(bmh *bmh_v1alpha1.BareMetalHost, age
 	return reconcileComplete{true}
 }
 
+// This function determines if the detached annotation should be added to the BMH
+// if the BMH is a SNO.
+//
+// The detached annotation is added if the installation of the agent associated with
+// the host has either started or has completed.
+func (r *BMACReconciler) updateBMHDetachedAnnotationIfSNO(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, agent *aiv1beta1.Agent) reconcileResult {
+
+	bmhAnnotations := bmh.ObjectMeta.GetAnnotations()
+	// Annotation already set to true
+	if bmhAnnotations[BMH_DETACHED_ANNOTATION] == "true" {
+		return reconcileComplete{}
+	}
+
+	// No updates if agent does not have a ClusterDeployment name
+	if agent.Spec.ClusterDeploymentName == nil {
+		return reconcileComplete{}
+	}
+	kubeKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      agent.Spec.ClusterDeploymentName.Name,
+	}
+
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	if err := r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		return reconcileError{err}
+	}
+
+	// If ClusterDeployment is not SNO do nothing
+	// A ClusterDeployment is SNO when ControlPlaneAgents == 1 and WorkerAgents == 0, see isSNO in
+	// https://github.com/openshift/assisted-service/blob/master/internal/controller/controllers/clusterdeployments_controller.go
+	if !(clusterDeployment.Spec.Provisioning.InstallStrategy.Agent.ProvisionRequirements.ControlPlaneAgents == 1 &&
+		clusterDeployment.Spec.Provisioning.InstallStrategy.Agent.ProvisionRequirements.WorkerAgents == 0) {
+		return reconcileComplete{}
+	}
+
+	if agent.Status.Conditions == nil {
+		return reconcileComplete{}
+	}
+	installConditionReason := conditionsv1.FindStatusCondition(agent.Status.Conditions, aiv1beta1.InstalledCondition).Reason
```

Comment:
An interesting point that I'm not sure how to resolve. After the installation is completed we remove the cluster https://github.com/openshift/assisted-service/blob/master/internal/controller/controllers/clusterdeployments_controller.go#L145
Right now we have a bug that we don't delete the agents after the installation is completed but we have a ticket to fix that, so once it is fixed it will not be long be before we finish the installation and deleting the host, so there may be a race that the agnet is already deleted before you get to mark BMH with this flag.

Reply:
i think we need to stop reconcile in such case no? we will set it already to detach while it will be installing. from the point agent will be deleted we will not reconcile on this bmh right? @flaper87 ?

Reply:
you will stop reconcile, a problem can be if you get the event late and the resource is already deleted, you will not get to this part at all.

Reply:
In this implementation, detached will be set also when the installation is in progress. I would expect that when the installation is complete, `detached` will be already set because it should have been said during the `InstallationInProgress` step.

Another (less elegant) option would be to not use the Agent conditions but to set/unset detached based on the following:

1. During the `reconcileBMH` step, where we know the BMH doesn't have the ISO URL set, therefore the installation cannot have started, we set `detached` to false: https://github.com/openshift/assisted-service/blob/480208622cd30a77922f3c2ef1948780b241f312/internal/controller/controllers/bmh_agent_controller.go#L350-L360
2. During the `agentInventory` reconcile, where we know the agent has started, we set `detached: true` so that Ironic stops the power management of this node: https://github.com/openshift/assisted-service/blob/480208622cd30a77922f3c2ef1948780b241f312/internal/controller/controllers/bmh_agent_controller.go#L245-L258


thoughts?

Reply:
Is there an option to unset it if the installation fails?

Reply:
> Is there an option to unset it if the installation fails?

In the current implementation, no. This is what I was after with the second proposal. We currently don't have a way to "restart" the installation unless the Image URL is removed from the BMH. Setting `detached: false` in the `reconcileBMH` function (right after we set the online status to True) would solve this issue: 

https://github.com/openshift/assisted-service/blob/480208622cd30a77922f3c2ef1948780b241f312/internal/controller/controllers/bmh_agent_controller.go#L398

Reply:
it's not related to restarting the installation, when users will try again he will create a new cluster deployment but because BMH will be detached it will not boot the new ISO right?

Reply:
> it's not related to restarting the installation, when users will try again he will create a new cluster deployment but because BMH will be detached it will not boot the new ISO right?

If we set `detached: false` as part of the `reconcileBMH` step (while we copy the ISOUrl to the BMH resource), I think we won't have this issue as BMO/ironic will manage that node at that point.

Reply:
I've updated reconcileBMH to remove the detached annotation if bmh.Spec.Image is removed.